File: cmd/arbos/voice.go
package main

import (
	"context"
	"crypto/sha256"
	_ "embed"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// voiceDictateSrc is the on-device dictation helper, compiled on first use.
//
//go:embed voice_dictate.swift
var voiceDictateSrc string

// voiceInfoPlist is the dictation app's Info.plist. The usage-description keys
// are what let macOS show the microphone / speech-recognition prompts; the app
// is launched via `open` so it is its own TCC-responsible process and these
// strings (not the launching terminal's) are read.
const voiceInfoPlist = `<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>CFBundleExecutable</key>
  <string>dictate</string>
  <key>CFBundleIdentifier</key>
  <string>com.unarbos.arbos.voice</string>
  <key>CFBundleName</key>
  <string>arbos voice</string>
  <key>CFBundlePackageType</key>
  <string>APPL</string>
  <key>CFBundleInfoDictionaryVersion</key>
  <string>6.0</string>
  <key>CFBundleVersion</key>
  <string>1</string>
  <key>CFBundleShortVersionString</key>
  <string>1.0</string>
  <key>LSUIElement</key>
  <true/>
  <key>NSMicrophoneUsageDescription</key>
  <string>arbos transcribes your speech into the chat composer.</string>
  <key>NSSpeechRecognitionUsageDescription</key>
  <string>arbos transcribes your speech into the chat composer.</string>
</dict>
</plist>
`

// voiceDonePoll is how often Stop checks for the helper's completion marker.
const voiceDonePoll = 100 * time.Millisecond

// voiceStopTimeout bounds how long Stop waits for the transcript after asking
// the helper to end — generous, since a first run may pause on the macOS
// permission prompt.
const voiceStopTimeout = 30 * time.Second

// hostVoice captures speech from the local machine's microphone and
// transcribes it on-device (Apple's Speech framework, via a tiny app bundle
// launched with `open`). One recording at a time — there is one mic. It
// satisfies gateway.VoiceRecorder.
type hostVoice struct {
	mu     sync.Mutex
	recDir string // per-recording control dir; "" when idle
	out    string // transcript path the helper writes
	stop   string // stop sentinel the helper watches
}

// voiceHelperApp compiles + signs the dictation app bundle once (keyed by a
// hash of its source) under the user cache dir, returning the .app path. It
// rebuilds only when the source or plist changes.
func voiceHelperApp() (string, error) {
	sum := sha256.Sum256([]byte(voiceDictateSrc + voiceInfoPlist))
	tag := hex.EncodeToString(sum[:8])
	cacheRoot, err := os.UserCacheDir()
	if err != nil {
		cacheRoot = os.TempDir()
	}
	dir := filepath.Join(cacheRoot, "arbos", "voice")
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return "", err
	}
	app := filepath.Join(dir, "arbos-dictate-"+tag+".app")
	bin := filepath.Join(app, "Contents", "MacOS", "dictate")
	if _, err := os.Stat(bin); err == nil {
		return app, nil
	}

	// Build into a temp app, then rename into place so a half-built bundle is
	// never observed.
	tmp := app + ".building"
	_ = os.RemoveAll(tmp)
	macos := filepath.Join(tmp, "Contents", "MacOS")
	if err := os.MkdirAll(macos, 0o755); err != nil {
		return "", err
	}
	src := filepath.Join(tmp, "dictate.swift")
	if err := os.WriteFile(src, []byte(voiceDictateSrc), 0o644); err != nil {
		return "", err
	}
	plist := filepath.Join(tmp, "Contents", "Info.plist")
	if err := os.WriteFile(plist, []byte(voiceInfoPlist), 0o644); err != nil {
		return "", err
	}

	ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
	defer cancel()
	if out, err := exec.CommandContext(ctx, "swiftc", "-O", "-o",
		filepath.Join(macos, "dictate"), src).CombinedOutput(); err != nil {
		return "", fmt.Errorf("compile voice helper: %w: %s", err, strings.TrimSpace(string(out)))
	}
	_ = os.Remove(src)
	// Ad-hoc sign so TCC binds the permission grant to a stable identity.
	if out, err := exec.CommandContext(ctx, "codesign", "--force", "--sign", "-", tmp).CombinedOutput(); err != nil {
		return "", fmt.Errorf("sign voice helper: %w: %s", err, strings.TrimSpace(string(out)))
	}
	if err := os.Rename(tmp, app); err != nil {
		return "", err
	}
	return app, nil
}

// Start begins capturing from the host microphone. The recording outlives the
// request that triggers it: the helper is an independent `open`-launched app,
// coordinated through files, and Stop ends it.
func (v *hostVoice) Start(_ context.Context) error {
	v.mu.Lock()
	defer v.mu.Unlock()
	if v.recDir != "" {
		return errors.New("a recording is already in progress")
	}
	app, err := voiceHelperApp()
	if err != nil {
		return err
	}
	recDir, err := os.MkdirTemp("", "arbos-voice-")
	if err != nil {
		return err
	}
	out := filepath.Join(recDir, "transcript.txt")
	stop := filepath.Join(recDir, "stop")
	// -n forces a fresh instance, -g keeps it from stealing focus.
	cmd := exec.Command("open", "-n", "-g", app, "--args", "--out", out, "--stop", stop)
	if combined, err := cmd.CombinedOutput(); err != nil {
		_ = os.RemoveAll(recDir)
		return fmt.Errorf("launch voice helper: %w: %s", err, strings.TrimSpace(string(combined)))
	}
	v.recDir, v.out, v.stop = recDir, out, stop
	return nil
}

// Stop ends the recording and returns the transcribed text: it drops the stop
// sentinel the helper watches, waits for the completion marker, then reads the
// transcript (or the error the helper recorded).
func (v *hostVoice) Stop(ctx context.Context) (string, error) {
	v.mu.Lock()
	recDir, out, stop := v.recDir, v.out, v.stop
	v.recDir, v.out, v.stop = "", "", ""
	v.mu.Unlock()

	if recDir == "" {
		return "", errors.New("no recording in progress")
	}
	defer func() { _ = os.RemoveAll(recDir) }()

	if err := os.WriteFile(stop, nil, 0o644); err != nil {
		return "", fmt.Errorf("signal voice helper: %w", err)
	}

	done := out + ".done"
	deadline := time.After(voiceStopTimeout)
	tick := time.NewTicker(voiceDonePoll)
	defer tick.Stop()
	for {
		if _, err := os.Stat(done); err == nil {
			break
		}
		select {
		case <-ctx.Done():
			return "", ctx.Err()
		case <-deadline:
			return "", errors.New("voice transcription timed out")
		case <-tick.C:
		}
	}

	if msg, err := os.ReadFile(out + ".err"); err == nil {
		return "", errors.New(strings.TrimSpace(string(msg)))
	}
	text, err := os.ReadFile(out)
	if err != nil {
		return "", fmt.Errorf("read transcript: %w", err)
	}
	return strings.TrimSpace(string(text)), nil
}

File: desktop/src/voice.rs
//! On-device dictation for the composer microphone.
//!
//! The mic is on this Mac — the window — not on a remote kernel. A tiny
//! signed app bundle talks to Apple Speech. Live partials land in a file.
//! Stop writes a sentinel and waits for the transcript.

use anyhow::{Context, Result, anyhow};
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::{Mutex, OnceLock};
use std::thread;
use std::time::{Duration, Instant};

#[cfg(target_os = "macos")]
use std::hash::{Hash, Hasher};
#[cfg(target_os = "macos")]
use std::process::Command;

const SWIFT: &str = include_str!("voice_dictate.swift");

const PLIST: &str = r#"<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>CFBundleExecutable</key>
  <string>dictate</string>
  <key>CFBundleIdentifier</key>
  <string>com.unarbos.arbos.voice</string>
  <key>CFBundleName</key>
  <string>arbos voice</string>
  <key>CFBundlePackageType</key>
  <string>APPL</string>
  <key>CFBundleInfoDictionaryVersion</key>
  <string>6.0</string>
  <key>CFBundleVersion</key>
  <string>1</string>
  <key>CFBundleShortVersionString</key>
  <string>1.0</string>
  <key>LSUIElement</key>
  <true/>
  <key>NSMicrophoneUsageDescription</key>
  <string>arbos transcribes your speech into the chat composer.</string>
  <key>NSSpeechRecognitionUsageDescription</key>
  <string>arbos transcribes your speech into the chat composer.</string>
</dict>
</plist>
"#;

const STOP_TIMEOUT: Duration = Duration::from_secs(30);
const DONE_POLL: Duration = Duration::from_millis(100);

struct Rec {
    dir: PathBuf,
    out: PathBuf,
    stop: PathBuf,
}

fn hold() -> &'static Mutex<Option<Rec>> {
    static HOLD: OnceLock<Mutex<Option<Rec>>> = OnceLock::new();
    HOLD.get_or_init(|| Mutex::new(None))
}

/// Begin capture. A leftover take from a crashed start is dropped first.
pub fn start() -> Result<()> {
    #[cfg(not(target_os = "macos"))]
    {
        Err(anyhow!("voice dictation is macOS-only for now"))
    }
    #[cfg(target_os = "macos")]
    {
        let mut hold = hold().lock().map_err(|_| anyhow!("voice lock poisoned"))?;
        if let Some(old) = hold.take() {
            abort(old);
        }
        *hold = Some(launch()?);
        Ok(())
    }
}

/// Latest partial text. Empty when idle or before the first word.
pub fn peek() -> Result<String> {
    let hold = hold().lock().map_err(|_| anyhow!("voice lock poisoned"))?;
    let Some(rec) = hold.as_ref() else {
        return Ok(String::new());
    };
    Ok(fs::read_to_string(&rec.out)
        .unwrap_or_default()
        .trim()
        .to_string())
}

/// End capture and return the transcript.
pub fn stop() -> Result<String> {
    let mut hold = hold().lock().map_err(|_| anyhow!("voice lock poisoned"))?;
    let Some(rec) = hold.take() else {
        return Err(anyhow!("no recording in progress"));
    };
    drop(hold);
    finish(rec)
}

#[cfg(target_os = "macos")]
fn launch() -> Result<Rec> {
    let app = helper_app()?;
    let dir = tempfile_dir()?;
    let out = dir.join("transcript.txt");
    let stop = dir.join("stop");
    let combined = Command::new("open")
        .args(["-n", "-g"])
        .arg(&app)
        .args(["--args", "--out"])
        .arg(&out)
        .arg("--stop")
        .arg(&stop)
        .output()
        .context("launch voice helper")?;
    if !combined.status.success() {
        let _ = fs::remove_dir_all(&dir);
        return Err(anyhow!(
            "launch voice helper: {}",
            String::from_utf8_lossy(&combined.stderr).trim()
        ));
    }
    Ok(Rec { dir, out, stop })
}

fn finish(rec: Rec) -> Result<String> {
    let _ = fs::write(&rec.stop, b"");
    let done = format!("{}.done", rec.out.display());
    let errp = format!("{}.err", rec.out.display());
    let deadline = Instant::now() + STOP_TIMEOUT;
    while !Path::new(&done).exists() {
        if Instant::now() >= deadline {
            let _ = fs::remove_dir_all(&rec.dir);
            return Err(anyhow!("voice transcription timed out"));
        }
        thread::sleep(DONE_POLL);
    }
    let err = fs::read_to_string(&errp).ok();
    let text = fs::read_to_string(&rec.out).unwrap_or_default();
    let _ = fs::remove_dir_all(&rec.dir);
    if let Some(msg) = err {
        let msg = msg.trim();
        if !msg.is_empty() {
            return Err(anyhow!("{msg}"));
        }
    }
    Ok(text.trim().to_string())
}

#[cfg(target_os = "macos")]
fn abort(rec: Rec) {
    let _ = fs::write(&rec.stop, b"");
    let done = format!("{}.done", rec.out.display());
    let deadline = Instant::now() + Duration::from_secs(2);
    while Instant::now() < deadline && !Path::new(&done).exists() {
        thread::sleep(Duration::from_millis(50));
    }
    let _ = fs::remove_dir_all(&rec.dir);
}

#[cfg(target_os = "macos")]
fn helper_app() -> Result<PathBuf> {
    let mut hasher = std::collections::hash_map::DefaultHasher::new();
    SWIFT.hash(&mut hasher);
    PLIST.hash(&mut hasher);
    let tag = format!("{:016x}", hasher.finish());
    let root = dirs::cache_dir()
        .unwrap_or_else(std::env::temp_dir)
        .join("arbos")
        .join("voice");
    fs::create_dir_all(&root).context("voice cache")?;
    let app = root.join(format!("arbos-dictate-{tag}.app"));
    let bin = app.join("Contents/MacOS/dictate");
    if bin.is_file() {
        return Ok(app);
    }

    let tmp = root.join(format!("arbos-dictate-{tag}.building.app"));
    let _ = fs::remove_dir_all(&tmp);
    let macos = tmp.join("Contents/MacOS");
    fs::create_dir_all(&macos).context("voice helper dirs")?;
    let src = tmp.join("dictate.swift");
    fs::write(&src, SWIFT).context("write voice helper source")?;
    fs::write(tmp.join("Contents/Info.plist"), PLIST).context("write voice helper plist")?;

    let compiled = Command::new("swiftc")
        .args(["-O", "-o"])
        .arg(macos.join("dictate"))
        .arg(&src)
        .output()
        .context("compile voice helper")?;
    if !compiled.status.success() {
        let _ = fs::remove_dir_all(&tmp);
        return Err(anyhow!(
            "compile voice helper: {}",
            String::from_utf8_lossy(&compiled.stderr).trim()
        ));
    }
    let _ = fs::remove_file(&src);

    let sign = Command::new("codesign")
        .args(["--force", "--sign", &signing_identity()])
        .arg(&tmp)
        .output()
        .context("sign voice helper")?;
    if !sign.status.success() {
        let _ = fs::remove_dir_all(&tmp);
        return Err(anyhow!(
            "sign voice helper: {}",
            String::from_utf8_lossy(&sign.stderr).trim()
        ));
    }
    fs::rename(&tmp, &app).context("install voice helper")?;
    Ok(app)
}

#[cfg(target_os = "macos")]
fn signing_identity() -> String {
    if let Ok(id) = std::env::var("ARBOS_SIGN_IDENTITY") {
        if !id.is_empty() {
            return id;
        }
    }
    let out = Command::new("security")
        .args(["find-identity", "-v", "-p", "codesigning"])
        .output();
    if let Ok(out) = out {
        if String::from_utf8_lossy(&out.stdout).contains("\"Arbos Dev\"") {
            return "Arbos Dev".into();
        }
    }
    "-".into()
}

#[cfg(target_os = "macos")]
fn tempfile_dir() -> Result<PathBuf> {
    let stamp = format!(
        "arbos-voice-{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_nanos())
            .unwrap_or(0)
    );
    let dir = std::env::temp_dir().join(stamp);
    fs::create_dir_all(&dir).context("voice rec dir")?;
    Ok(dir)
}
